In [42]:
# Import required libraries
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [43]:
# Load the cryptocurrency market data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",  # Path to the CSV file
    index_col="coin_id"                 # Set the 'coin_id' column as the index
)

# Display the first 10 rows of the DataFrame
df_market_data.head(10)


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [12]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [44]:
# Plot a line chart of the data to visualize trends
df_market_data.hvplot.line(
    width=800,  # Set plot width
    height=400, # Set plot height
    rot=90      # Rotate x-axis labels by 90 degrees
)


:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [45]:
# Standardize the data using StandardScaler
scaler = StandardScaler()
scaled_data = scaler.fit_transform(df_market_data)  # Normalize numerical values

In [46]:
# Convert the scaled data back into a Pandas DataFrame
df_market_data_scaled = pd.DataFrame(
    scaled_data,                       # Use scaled data
    columns=df_market_data.columns     # Set original column names
)

# Add the original 'coin_id' index back to the DataFrame
df_market_data_scaled["coin_id"] = df_market_data.index

# Set 'coin_id' as the index of the new DataFrame
df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display the first five rows of the scaled DataFrame
df_market_data_scaled.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [47]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12)) 

In [48]:
# Initialize an empty list to store inertia values
inertia = []

# Compute the inertia for each k-value
for k in k_values:
    # Create a KMeans model with the current k
    k_model = KMeans(n_clusters=k, random_state=0)
    
    # Fit the model to the scaled data
    k_model.fit(df_market_data_scaled)
    
    # Append the inertia value to the list
    inertia.append(k_model.inertia_)


In [49]:
# Create a dictionary with k-values and corresponding inertia values
elbow_data = {"k": k_values, "inertia": inertia}

# Convert the dictionary into a Pandas DataFrame
df_elbow = pd.DataFrame(elbow_data)


In [50]:
# Plot the Elbow Curve to identify the optimal number of clusters
df_elbow.hvplot.line(
    x="k",           # x-axis: k-values
    y="inertia",     # y-axis: inertia values
    title="Elbow Curve",  # Title for the plot
    xticks=k_values      # Set x-axis ticks to match k-values
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [52]:
# Set the best k-value based on the Elbow Curve
best_k = 4

# Create a KMeans model with the optimal number of clusters
model = KMeans(n_clusters=best_k, random_state=0)

In [53]:
# Fit the K-Means model to the scaled data
model.fit(df_market_data_scaled)

KMeans(n_clusters=4, random_state=0)

In [54]:
# Predict the cluster labels for the scaled data
predicted_clusters = model.predict(df_market_data_scaled)

# Print the resulting array of cluster labels
print(predicted_clusters)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [23]:
# Create a copy of the DataFrame
df_market_data_predictions = df_market_data.copy()


In [55]:
# Add a new column to the DataFrame with the predicted clusters
df_market_data_predictions["predicted_cluster"] = predicted_clusters

# Display sample data
df_market_data_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                           1.08388                     7.60278   
ethereum                          0.22392                    10.38134   
tether                           -0.21173                     0.04935   
ripple                           -0.37819                    -0.60926   
bitcoin-cash                      2.90585                    17.09717   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                           6.57509                      7.67258   
ethereum                          4.80849                      0.13169   
tether                            0.00640                     -0.04237   
ripple                            2.24984                      0.23455   
bitcoin-cash                     14.75334                     15.74903   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                          -3.25185                      83.51840   
ethereum                        -12.88890                     186.77418   
tether                            0.28037                      -0.00542   
ripple                          -17.55245                      39.53888   
bitcoin-cash                    -13.71793                      21.66042   

              price_change_percentage_1y  predicted_cluster  
coin_id                                                      
bitcoin                         37.51761                  2  
ethereum                       101.96023                  2  
tether                           0.01954                  0  
ripple                         -16.60193                  0  
bitcoin-cash                    14.49384                  2

In [56]:
# Create a scatter plot to visualize the clusters based on original data
df_market_data_predictions.hvplot.scatter(
    x="price_change_percentage_24h",  # x-axis: 24-hour price change
    y="price_change_percentage_7d",   # y-axis: 7-day price change
    by="predicted_cluster",           # Color points by cluster
    hover_cols=["coin_id"],           # Show 'coin_id' on hover
    title="Cryptocurrency Clusters: 24h vs. 7d Price Change"
)


:NdOverlay   [predicted_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [57]:
# Create a PCA model to reduce the data to three components
pca = PCA(n_components=3)

In [58]:
# Fit and transform the scaled data using PCA
pca_data = pca.fit_transform(df_market_data_scaled)

# Create a new DataFrame with the PCA-transformed data
df_market_pca = pd.DataFrame(
    data=pca_data,                   # PCA-transformed data
    columns=["PC1", "PC2", "PC3"],   # Principal component labels
    index=df_market_data.index       # Preserve original index
)

# Display the first five rows of the PCA DataFrame
df_market_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

In [29]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_
print(f"Explained variance of each component: {explained_variance}")
print(f"Total explained variance: {explained_variance.sum()}")


Explained variance of each component: [0.3719856  0.34700813 0.17603793]
Total explained variance: 0.8950316570309842


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.895 or 89.5%

In [30]:
# Create a new DataFrame with the PCA data.
df_market_data_pca = pd.DataFrame(
    pca_data, columns=["PC1", "PC2", "PC3"], index=df_market_data.index
)

# Copy the crypto names from the original data
df_market_data_pca["coin_id"] = df_market_data.index

# Set the coinid column as index
df_market_data_pca = df_market_data_pca.set_index("coin_id")

# Display sample data
df_market_data_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [31]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))

In [32]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=k, random_state=0)
    
    # 2. Fit the model to the data using `df_market_data_pca`
    k_model.fit(df_market_data_pca)
    
    # 3. Append the model.inertia_ to the inertia list
    inertia.append(k_model.inertia_)
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k": k_values, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)


In [33]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k": k_values, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)


In [60]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve (PCA Data)", 
    xticks=k_values
)


:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:**


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** 

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [35]:
# Initialize the K-Means model using the best value for k
best_k_pca = 4  # Based on the elbow curve in the graph

model_pca = KMeans(n_clusters=best_k_pca, random_state=0)


In [36]:
# Fit the K-Means model using the PCA data
model_pca.fit(df_market_data_pca)

KMeans(n_clusters=4, random_state=0)

In [37]:
# Predict the clusters to group the cryptocurrencies using the PCA data
predicted_clusters_pca = model_pca.predict(df_market_data_pca)

# Print the resulting array of cluster values.
predicted_clusters_pca

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

In [38]:
# Create a copy of the DataFrame with the PCA data
df_market_data_pca_predictions = df_market_data_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_market_data_pca_predictions['pca_cluster'] = predicted_clusters_pca

# Display sample data
df_market_data_pca_predictions.head()

PC1       PC2       PC3  pca_cluster
coin_id                                                
bitcoin      -0.600667  0.842760  0.461595            2
ethereum     -0.458261  0.458466  0.952877            2
tether       -0.433070 -0.168126 -0.641752            0
ripple       -0.471835 -0.222660 -0.479053            0
bitcoin-cash -1.157800  2.041209  1.859715            2

In [39]:
df_market_data_pca_predictions.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="pca_cluster",  # Color by PCA cluster
    hover_cols=["coin_id"],  # Show coin_id on hover
    title="Cryptocurrency Clusters (PCA): PC1 vs. PC2"
)


:NdOverlay   [pca_cluster]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [40]:
elbow_curve_original = df_elbow.hvplot.line(x='k', y='inertia', label='Original Data')
elbow_curve_pca = df_elbow_pca.hvplot.line(x='k', y='inertia', label='PCA Data')

(elbow_curve_original * elbow_curve_pca).opts(
    title="Elbow Curve Comparison (Original vs. PCA)", 
    legend_position='top_right'
)


:Overlay
   .Curve.Original_Data :Curve   [k]   (inertia)
   .Curve.PCA_Data      :Curve   [k]   (inertia)

In [41]:
# Create scatter plot for original data
original_scatter = df_market_data_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="predicted_cluster",
    hover_cols=["coin_id"],
    title="Original Clusters"
)

# Create scatter plot for PCA data
pca_scatter = df_market_data_pca_predictions.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="pca_cluster",
    hover_cols=["coin_id"],
    title="PCA Clusters"
)

# Display plots side by side
layout = (original_scatter + pca_scatter).cols(1)
layout


:Layout
   .NdOverlay.I  :NdOverlay   [predicted_cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [pca_cluster]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 